webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts

```diff
@@ -0,0 +1,203 @@
+    }
+}
+
+/**
+ * An IntervalDataLoader is an object that can be used to batch requests for fetching objects from the server. Instead
+ * of requesting data immediately, it will periodically check if any objects need to be requested from the server.
+ *
+ * It's intended to be used for loading low priority data such as information needed in response to WebSocket messages
+ * that the user won't see immediately.
+ */
+export class IntervalDataLoader<Identifier, Result = unknown> extends DataLoader<Identifier, Result> {
+    private intervalId: number = -1;
+
+    public startIntervalIfNeeded(ms: number): void {
+        if (this.intervalId !== -1) {
+            return;
+        }
+
+        this.intervalId = window.setInterval(() => this.fetchBatchNow(), ms);
+    }
+
+    public stopInterval(): void {
+        clearInterval(this.intervalId);
+        this.intervalId = -1;
+    }
+
+    public fetchBatchNow(): void {
+        const {identifiers} = this.startBatch();
+
+        if (identifiers.length === 0) {
+            return;
+        }
+
+        this.fetchBatch(identifiers);
+    }
+
+    public isBusy(): boolean {
+        return super.isBusy() || this.intervalId !== -1;
+    }
+}
+
+/**
+ * A DelayedDataLoader is an object that can be used to batch requests for fetching objects from the server. Instead of
+ * requesting data immediately, it will wait for an amount of time and then send a request to the server for all of
+ * the data which would've been requested during that time.
+ *
+ * More specifically, when queueForLoading is first called, a timer will be started. Until that timer expires, any other
+ * calls to queueForLoading will have the provided identifiers added to the ones from the initial call. When the timer
+ * finally expires, the request will be sent to the server to fetch that data. After that, the timer will be reset and
+ * the next call to queueForLoading will start a new one.
+ *
+ * DelayedDataLoader is intended to be used for loading data for components which are unaware of each other and may appear
+ * in different places in the UI from each other which could otherwise send repeated requests for the same or similar
+ * data as one another.
+ */
+export class DelayedDataLoader<Identifier> extends DataLoader<Identifier, Promise<unknown>> {
+    private readonly wait: number = -1;
+
+    private timeoutId: number = -1;
+    private timeoutCallbacks = new Set<{
+        identifiers: Set<Identifier>;
+        resolve(): void;
+    }>();
+
+    constructor(args: {
+        fetchBatch: (identifiers: Identifier[]) => Promise<unknown>;
+        maxBatchSize: number;
+        wait: number;
+    }) {
+        super(args);
+
+        this.wait = args.wait;
+    }
+
+    public queueForLoading(identifiersToLoad: Identifier[]): void {
+        super.queueForLoading(identifiersToLoad);
+
+        this.startTimeoutIfNeeded();
+    }
+
+    public queueAndWait(identifiersToLoad: Identifier[]): Promise<void> {
```

Comment:
Waiting feels like the most complicated part of the data loader, and seem to just entrench the existing data model practices instead of moving us away from that temporal coupling.

Building on your comment above re: most places don't wait for results, could we explore removing the need for callers to wait and solving the unit testing another way?

Reply:
I do agree that removing that would be ideal, but apart from figuring out how to test this (something like Testing Library's `waitFor` seems like it would work), I think the biggest hurdle for that would be to make the code more resilient to missing information since one place we tend to wait for things like this is when we need related information for an object (like the author of a post). I can't really see how we'd get around that without some sort of synchronization.

This doesn't seem like something we can solve right now, but I'll look at making it more explicit if the caller is awaiting on the result like with `queueToLoad` vs `queueAndWait`. That might lead to code that's a bit repetitive, but it'll at least help differentiate those two cases enough that people should see more usage of the non-blocking `queueToLoad`

Reply:
> I think the biggest hurdle for that would be to make the code more resilient to missing information since one place we tend to wait for things like this is when we need related information for an object (like the author of a post).

@hmhealey, can you point me an example? At least for components, I was wondering if it could just handle the partial state, with the React component refreshing after the partial update to potentially trigger the second load. For non-component contexts, like dispatches, I wonder if we can get the information we need in a single API request instead?

Reply:
The places that we await on this for something other than an optimization are in a couple places in the threads actions:
- https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/packages/mattermost-redux/src/actions/threads.ts#L71
- https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/packages/mattermost-redux/src/actions/threads.ts#L141

In general though, there's a lot of places where we use this sort of "wait for new data" throughout a lot of the initial load and routing logic. A lot of that comes down a lot to the lack of data ownership and the overfetching that that requires like I documented in [that spec I wrote last year](https://docs.google.com/document/d/1jLFhV2P7NGafCYfR3UcX6z4JWBZEj4IKrl4nIxRJK3c/edit), so it'll be hard to remove the waiting until we have more infrastructure like `DataLoader` in place. Once code can load data more easily and more automatically, it'll be easier for it to wait for Redux to be populated rather than explicitly waiting on an action like it has has to do now.

As for getting more data in a single API request, I've proposed that in the past as "richer APIs", and GraphQL was moving in that direction as well, but that goes against the focus on more RESTful APIs that we've had since API v4.

All that said, I'm going to leave this as-is since, while supporting waiting is more complicated, the current code in this PR should be able to handle that as long as it has time to soak. Not waiting would be ideal, but it requires rethinking a lot of how we do things, and we need to try this out before we can commit to it